### DISCUSSION TITLE
Provide dedicated virtual "signal-response" methods on objects

### DISCUSSION THREAD

**Discussion by mieldepoche** [godotengine/godot-proposals#4793] [Scripting]
Quite often I need to connect signals on an object's own script:

```gdscript
# in godot 4.0
extends Area3D

func _ready() -> void:
	body_entered.connect(_on_body_entered)
```

Taking inspiration from the dedicated notification response (`_process`, `_physics_process`...), what about providing virtual methods that if defined are automatically connected to their signal?

1. on each Object-derived class, for each `{signal}`, add a method named `_on_{signal}`
2. on `_init`, connect signals to overriden methods (with default connection flags):
```gdscript
# GDScript pseudocode
func _init() -> void:
    for sgl in get_signal_list():
        var meth = "_on_" + sgl.name
        if has_method(meth):
            connect(sgl.name, Callable(self, meth))
```
3. `_on_{user signal}` *could* be detected and connected too maybe, but providing autocompletion for inexistent methods is another problem
4. Such methods could have an icon next to them in the script editor indicating that they are "signal receivers" and will be automatically connected, like the icon displayed when connecting using the editor dialog: 
![connect icon](https://user-images.githubusercontent.com/71937306/176926226-1682c5a1-44a2-42f1-bf81-062e7207eff5.png)

Benefits:

1. an intuitive behavior similar to the ubiquitous `_process` et al.: provide a simpler-than-the-generic-case interface which enables itself upon usage

Limitations:
1. very convenient for simple cases, but would not support custom argument binds

While I felt the need for such a feature several times, I can't decide if it makes actual sense.
What do you think?

**Comment by xix-xeaon** [comment]
I agree that it's annoying to write the connections but this is a bit too automagic I think, I'd still like signal connections to be clearly defined. But there could be much better syntax sugar, like using annotations to connect functions:
```gdscript
@connect(body_entered)
func _on_body_entered(body):
    pass
```
This would essentially add the appropriate connect code to "_ready". This should work for all the features of signal-connections. You could refer to a signal elsewhere in the tree and bind more variables etc. If you want to support ultimate laziness then an "empty" connect-annotation could infer the signal from the function-name by stripping the "\_on\_"-prefix.
